src/o365/spo/commands/tenant/tenant-recyclebinitem-restore.ts

```diff
@@ -0,0 +1,203 @@
+import SpoCommand from '../../../base/SpoCommand';
+import GlobalOptions from '../../../../GlobalOptions';
+import { CommandOption, CommandCancel, CommandValidate } from '../../../../Command';
+
+const vorpal: Vorpal = require('../../../../vorpal-init');
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  url: string;
+  wait?: boolean;
+}
+
+interface RestSpoOperation {
+  HasTimedout: boolean;
+  IsComplete: boolean;
+  PollingInterval: number;
+}
+
+class SpoTenantRecycleBinItemRestoreCommand extends SpoCommand {
+  private timeout?: NodeJS.Timer;
+  private readonly maxAttempts: number = 5;
+
+  public get name(): string {
+    return commands.TENANT_RECYCLEBINITEM_RESTORE;
+  }
+
+  public get description(): string {
+    return 'Restores the specified deleted site collection from tenant recycle bin';
+  }
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.url = typeof args.options.url !== 'undefined';
+    telemetryProps.wait = typeof args.options.wait !== 'undefined';
+    return telemetryProps;
+  }
+
+  public commandAction(cmd: CommandInstance, args: CommandArgs, cb: (err?: any) => void): void {
+    const wait: boolean = args.options.wait || false;
+    let spoAdminUrl: string;
+
+    this.getSpoAdminUrl(cmd, this.debug)
+    .then((adminUrl: string): Promise<RestSpoOperation> => {
+      spoAdminUrl = adminUrl;
+      return this.restoreDeletedSite(args.options.url, spoAdminUrl, cmd);
+    })
+    .then((response: RestSpoOperation): Promise<void> => {
+      if (!response.HasTimedout && response.IsComplete) {
+        if (this.verbose) {
+          cmd.log('site collection restored');
+        }
+
+        return Promise.resolve();
+      }
+      else if (wait) {
+        return new Promise<void>((resolve: () => void, reject: (error: any) => void): void => {
+          this.waitForRestoreDeletedSite(
+            cmd,
+            this,
+            spoAdminUrl,
+            args.options.url,
+            resolve,
+            reject,
+            0
+          );
+        });
+      }
+      else {
+        return Promise.reject('site collection has not been restored');
+      }
+    })
+    .then(() => {
+      if (this.verbose) {
+        cmd.log(vorpal.chalk.green('DONE'));
+      }
+
+      cb()
+    }, (err: any): void => this.handleRejectedODataJsonPromise(err, cmd, cb));
+  }
+
+  public cancel(): CommandCancel {
+    return (): void => {
+      if (this.timeout) {
+        clearTimeout(this.timeout);
+      }
+    }
+  }
+
+  private waitForRestoreDeletedSite(cmd: CommandInstance, command: SpoTenantRecycleBinItemRestoreCommand, spoAdminUrl: string, siteToRestoreUrl: string, resolve: () => void, reject: (error: any) => void, iteration: number): void {
+    iteration++;
+
+    new Promise((): Promise<void> => {
+      return this.restoreDeletedSite(siteToRestoreUrl, spoAdminUrl, cmd)
```

Comment:
Were you able to verify that this is the right approach to wait for the site to be deleted? From what I've seen: typically, each site is restored before the first call finishes so the `waitForRestoreDeletedSite` method is never used. The `RestoreDeletedSite` API does return an `SpoOperation` which indicates that theoretically the operation could span beyond a single call. I thought of mimicking what the behavior would be if this method was called by executing restore of the same site twice in parallel. In that case, one of the commands executes as expected while the other one fails with a `400` response error. To me it seems like requesting restore of a site that's already being restored is not the right way to check if the operation is still in progress. If I look at how that check is done in CSOM, you would check the value of the `IsComplete` property and update the `SpoOperation` object after the specified polling interval to get its latest status. The problem with the REST API is that it doesn't return the object ID so there is no way for us to get updated operation status. With that I wonder if we shouldn't switch to using CSOM after all, which does allow us to poll for update (see sample implementation in `spo site classic add` for example).

@garrytrinder @VelinGeorgiev @appieschot have you by any chance used the `RestoreDeletedSite` REST API in any of your projects to shed some more light on using it correctly?

Reply:
I haven't. I am sorry

Reply:
Same, I’ve not got experience of using this endpoint, sorry.

Reply:
Nice try about mimicking the behavior 👍  
I haven't though about that, but only "guessing", which is not quite enough I agree. 

About `SpoOperation` object, that was my first approach when thinking about the implementation: I was also hoping that I could get some info about "progression" from the endpoint, like `JobState` or `_ObjectIdentity_` in client.svc. But I haven't found anything so I created months ago an issue [here](https://github.com/SharePoint/sp-dev-docs/issues/5368) which hasn't been answered since. I've also noticed that a similar approach is made in the PnP-Sites-Core library, for site creation [here](https://github.com/pnp/PnP-Sites-Core/blob/28aa3c8f194dc6dac4bdbf54f9b93648c1aa4f69/Core/OfficeDevPnP.Core/Extensions/TenantExtensions.cs#L144).

I'm not against the idea to use CSOM, but it would be great to have information about this object and how it behaves in REST API, because it's used in other endpoints such as site creation, remove deleted site collection and more 🙂

Reply:
I've seen your issue indeed. The trouble with the REST API is that it's not officially supported so in case of any issue, the first suggestion is to go with CSOM. In our case, we could get the job state from CSOM but since the job ID is not exposed in REST, it would mean that we need to refactor the whole command to CSOM.

Would you prefer to hold off a bit and see if we can get a response from Microsoft or would you prefer to get going and refactor the command to CSOM @michaelmaillot?

Reply:
Both...? 😬 

I would like to have an answer from Microsoft indeed. But in the meantime, I'm going to have a look about refactoring the command in CSOM, based on existing commands.

But I'm curious. What do you mean by 

> The trouble with the REST API is that it's not officially supported

Supported by the CLI community ? I perfectly understand.
Supported by Microsoft ? As it's continuously updated with new methods and properties, how can this cannot be supported ? Have I miss something about the "preferred" endpoint URI usage ?

However, I understand that the CSOM (mean using `client.svc` approach) is better for some operations as you can get status info about them.

Reply:
Microsoft's stance is that if an API is not documented, it's not officially supported for third-party use. Meaning, if you run into any trouble, you're on your own. The only REST APIs that are documented are the basic list operations so in this particular case, we can't really report it as an issue other than an inconvenience and perhaps a UserVoice entry asking Microsoft to fill the gap.